### ISSUE TITLE
xDS gRPC credentials for AWS IAM

### ISSUE THREAD

**Issue by lavignes** [envoyproxy/envoy#5215]
*Title*: xDS gRPC credentials for AWS IAM

*Description*:
I work on the AWS App Mesh team. For our public preview we added a [core.GrpcService.GoogleGrpc.CallCredentials][1] plugin that uses a slightly modified [signature v4][2] signing process to authenticate xDS requests. Essentially, we map the gRPC POST request into an HTTP 1.1-style request and pass follow the signing algorithm as documented.

We'd like to contribute this plugin into upstream Envoy to allow users to easily make their own builds and  get timely feature updates/bugfixes etc.

Our current implementation pulls in a (opt-in) dependency on the [core library of AWS C++ SDK][3], which is probably overkill. It also relies on linking the SDK against openssl and libcurl to perform the necessary HTTP requests to fetch credentials from places like the [EC2 instance metadata endpoint][4].

I was exploring dropping the AWS SDK dependency and passing a context object to the `Envoy::Grpc::GoogleGrpcCredentialsFactory` with a cluster manager so that we could leverage Envoy's HTTP apis to call to the metadata endpoints, but I'm worried I'm going down the completely wrong path to making REST api calls from within the CredentialsFactory.

Any suggestions? I'm happy to open a PR with the current implementation, but I'm sure a minimalist AWS-auth lib in Envoy could be handy for other extensions that may need to communicate with AWS services. ex: #4526

[1]: https://www.envoyproxy.io/docs/envoy/latest/api-v2/api/v2/core/grpc_service.proto#core-grpcservice-googlegrpc-callcredentials
[2]: https://docs.aws.amazon.com/general/latest/gr/signature-version-4.html
[3]: https://github.com/aws/aws-sdk-cpp/tree/master/aws-cpp-sdk-core
[4]: https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-instance-metadata.html

**Comment by mattklein123** [comment]
Thanks @lavignes this sounds very useful. @htuch and @lizan are most familiar with the gRPC extensions and can hopefully provide some guidance here.

**Comment by dio** [comment]
This is interesting since if we go by extending `grpc::MetadataCredentialsPlugin` we need to find a way of `GetMetadata` from the EC2 instance metadata endpoint within the google gRPC thread (which requires the opt-in aws-sdk for now? Do we have a way to do that without it?).

**Comment by lavignes** [comment]
@dio
> requires the opt-in aws-sdk for now?  Do we have a way to do that without it?

We added the aws-sdk dependency behind a compiler-flag, e.g. `--enable=aws_sdk` We only really use it to curl the credentials endpoint and sign. Both are pretty trivial to implement without the SDK, granted we can safely make the instance metadata request.

I believe I could dynamically register the metadata endpoint as a remote cluster, thereby allowing us to make GET calls to the endpoint from within the plugin. But don't know if that's the right way to proceed. Especially since I do need to block on that GET request as well.

**Comment by htuch** [comment]
@lavignes I think you could do it in `Envoy::Grpc::GoogleGrpcCredentialsFactory` with some changes. It would need to become an asynchronous API, rather than its current synchronous interface, as it will need to wait for HTTP completions before continuing. 

If you are executing from within the `grpc::MetadataCredentialsPlugin()` execution flow, you have a lot more freedom when it comes to threading; those plugins can block as the gRPC core runtime can schedule you on another thread. I don't think you can easily resuse Envoy's cluster manager there, as cluster manager is designed around a notion of a fixed number of worker threads, all of which are owned by Envoy and have updates coordinated in a safe fashion. They can't be used on gRPC threads.

If the gRPC libs don't provide an HTTP fetch method, I would suggest a number of options:
1. Fetch in `Envoy::Grpc::GoogleGrpcCredentialsFactory()` (with consequent changes to the interfaces for this).
2. Use libcurl or whatever inside `grpc::MetadataCredentialsPlugin()`. Add this as an external dependency to Envoy.
3. Build your own HTTP client from Envoy components and use it in `grpc::MetadataCredentialsPlugin()`, with no reuse of the existing notion of clusters or cluster manager or connection pools. You can see an example of how this is done by looking at the health checker code in `HealthCheckerImplBase`.

I think (3) could be the most Envoy-ish way of approaching this problem, (1) would be next preference. (2) seems like we're bringing in an unnecessary new dep.

**Comment by lavignes** [comment]
@htuch Thanks! This is exactly what I was looking for. We'll go ahead with implementing (3).

**Comment by mattklein123** [comment]
@htuch @lavignes one other potential thing to consider would be to not use the Google gRPC client at all. Presumably you are just using it because it has a credentials/auth hook? Envoy's built-in AsyncClient/GrpcAsyncClient I think could be fairly easily extended to support filters, and then you could have an AWS auth filter plugin of some kind (or just a metadata callback which does the right thing). I'm not sure if this is easier or harder than (3) but just throwing it out there.

**Comment by htuch** [comment]
@mattklein123 yeah, this would be totally reasonable. I guess at some point we will need to extend the Envoy gRPC client to support a wider range of auth models. This would have been a place that it would make sense to try and leverage libraries from Google gRPC perhaps, but I think it needs a more detailed investigation to figure out if these APIs exist today and/or coordinate with gRPC team to expose them for reuse in Envoy gRPC.

**Comment by lavignes** [comment]
This is on-hold right now since I'm waiting on #5387 to add the libcurl dependency (as mentioned [here](https://github.com/envoyproxy/envoy/pull/5546#issuecomment-452568475)).

**Comment by lizan** [comment]
@lavignes is libcurl dependency is your only blocker? I think we may extract that out from #5387 if that is going to be on hold for a while.

**Comment by lavignes** [comment]
@lizan Yeah. I'm fine with that. I'll just base off of the curl dependency in #5387. I'll try and send out another pull request next week.

**Comment by lavignes** [comment]
Sorry for the really long delay here. I'm going to getting back on this soon. Most of the code is already written on my branch, but isn't quite ready for a PR

**Comment by dio** [comment]
@lavignes anything we can do to help? If you're OK probably we can do the PR on your behalf.

**Comment by ivitjuk** [comment]
I spoke with @lavignes and the plan is that I take over this issue. After going through the thread I think following would be a way to go:

- a separate PR with only libcurl added as a new dependency (extracted from #5387)
- a set of new PRs based on #5546 but with reduced scope and based on libcurl

If that makes sense, I'll move forward with the implementation.

**Comment by dio** [comment]
@ivitjuk very much appreciated. Yes, please.

**Comment by lizan** [comment]
Sounds good to me, thanks!

**Comment by lavignes** [comment]
Bumping to keep this alive. Its almost done! Pending #7303 and a final PR dependent on it.

**Comment by lavignes** [comment]
Not stale. Pending #7532
